compiler/src/iree/compiler/Codegen/WGSL/WGSLReplacePushConstants.cpp

```diff
@@ -0,0 +1,137 @@
+#include "iree/compiler/Dialect/Flow/IR/FlowOps.h"
+#include "iree/compiler/Dialect/HAL/IR/HALDialect.h"
+#include "iree/compiler/Dialect/HAL/IR/HALOps.h"
+#include "mlir/Dialect/Arithmetic/IR/Arithmetic.h"
+#include "mlir/Dialect/Tensor/IR/Tensor.h"
+#include "mlir/IR/Attributes.h"
+#include "mlir/IR/Builders.h"
+#include "mlir/Pass/Pass.h"
+
+namespace mlir {
+namespace iree_compiler {
+namespace {
+
+// These must match what the runtime uses.
+#define IREE_HAL_WEBGPU_PARAMS_BIND_GROUP_INDEX 3
+#define IREE_HAL_WEBGPU_PARAMS_BINDING_INDEX 0
+
+static void replaceConstantLoadOp(IREE::Flow::DispatchTensorLoadOp loadOp,
+                                  IREE::HAL::InterfaceConstantLoadOp op) {
+  OpBuilder builder(op);
+
+  // tensor.extract -> i32
+  auto offsetValue = builder.createOrFold<arith::ConstantIndexOp>(
+      op.getLoc(), op.getIndex().getZExtValue());
+  auto extractOp =
+      builder.create<tensor::ExtractOp>(op.getLoc(), loadOp, offsetValue);
+
+  // Convert to original type.
+  auto opType = op.getType();
+  if (opType.isIndex()) {
+    auto indexCastOp =
+        builder.create<arith::IndexCastOp>(op.getLoc(), opType, extractOp);
+    op.replaceAllUsesWith(indexCastOp.getResult());
+  } else {
+    // TODO(scotttodd): truncate or zero-extend types with different bitwidths?
+    auto bitcastOp =
+        builder.create<arith::BitcastOp>(op.getLoc(), op.getType(), extractOp);
+    op.replaceAllUsesWith(bitcastOp.getResult());
+  }
+
+  op.erase();
+}
+
+class WGSLReplacePushConstantsPass
+    : public WGSLReplacePushConstantsBase<WGSLReplacePushConstantsPass> {
+  void getDependentDialects(DialectRegistry &registry) const override {
+    registry.insert<mlir::arith::ArithmeticDialect, mlir::func::FuncDialect,
+                    mlir::tensor::TensorDialect, IREE::Flow::FlowDialect,
+                    IREE::HAL::HALDialect>();
+  }
+
+  void runOnOperation() override {
+    auto parentOp = getOperation();
+    auto loc = parentOp.getLoc();
+    auto constantLoadOps = llvm::to_vector<4>(
+        parentOp.getOps<IREE::HAL::InterfaceConstantLoadOp>());
+    if (constantLoadOps.empty()) return;
+
+    OpBuilder builder(parentOp);
+    builder.setInsertionPointToStart(&parentOp.getBlocks().front());
+
+    // Group all push constants into a single `hal.interface.binding.subspan`
+    // and load from it once using `flow.dispatch.tensor.load`, then extract
+    // individual push constants with `tensor.extract`.
+
+    // Find the range of push constant indices (0 to some maximum).
+    uint64_t maxConstantIndex = 0;
+    // Inspect the alignment values. These are just hints, so if all are equal
+    // then use the value, otherwise drop the alignment hint.
+    SmallVector<uint64_t> alignmentValues;
+    bool missingAlignmentValue = false;
+    for (auto constantLoadOp : constantLoadOps) {
+      maxConstantIndex =
+          std::max(constantLoadOp.getIndex().getZExtValue(), maxConstantIndex);
```

Comment:
hmmm, this can get tricky. If you pick this as the overall alignment, then within the packed constant arrays you will need to probably create holes to retain the same alignment.

I understand this is a hint. I'd expect things to be less surprising if you just drop the alignment for all the individual constant loads. You can have an alignment for the binding you are creating which is completely decoupled.

Reply:
Interesting point. This isn't actually _packing_ constants though (but it could?). The dispatch tensor is Nxi32, with one element for each push constant. When values are extracted from each i32 slot they should be bitcast and truncated or zero extended to that expected type (currently I just have this bitcasting).

So if the truncation was implemented, how should alignment work with i16 values, for example? (I'm not sure right now)

Reply:
I think the shared alignment still works at other bit depths - as long as the values are not packed together and each uses a unique i32 slot in this buffer.

Reply:
you would model it the same way a regular C struct is modeled in LLVM. I dont think there is a fixed spec for it in the language standard but say you have this
```
struct args {
  int16 v1;
  int32 v2;
};
```

Typically the struct above is of size 8 bytes with a hole of 2 bytes between `v1` and `v2`. since `v2` is 4 byte aligned. If everything we are getting is i32 values, then its all good.